internal/ensure/requiredversion.go

```diff
@@ -0,0 +1,124 @@
+	"golang.org/x/tools/go/packages"
+)
+
+const (
+	orchestrionPkgPath = "github.com/datadog/orchestrion"
+	envVarRespawned    = "DD_ORCHESTRION_RESPAWNED_FOR"
+	envValRespawnLocal = "<local>"
+)
+
+var errRespawnLoop = errors.New("re-spawn loop detected")
+
+// RequiredVersion makes sure the version of the tool currently running is the same as the one
+// required in the current working directory's "go.mod" file by calling `syscall.Exec` with the
+// relevant `go run` command if necessary to replace the current process with one using the required
+// version.
+//
+// If this returns `nil`, the current process is running the correct version of the tool and can
+// proceed with it's intended purpose. If it returns an error, that should be presented to the user
+// before exiting with a non-0 status code. If the process was correctly substituted, this function
+// never returns control to its caller (as the process has been replaced).
+func RequiredVersion() error {
+	return requiredVersion(goModVersion, os.Getenv, syscall.Exec, os.Args)
+}
+
+// requiredVersion is the internal implementation of RequiredVersion, and takes the goModVersion and
+// syscall.Exec functions as arguments to allow for easier testing.
+func requiredVersion(
+	goModVersion func() (string, error),
+	osGetenv func(string) string,
+	syscallExec func(argv0 string, argv []string, env []string) error,
+	osArgs []string,
+) error {
+	required, err := goModVersion()
+	if err != nil {
+		return fmt.Errorf("failed to determine go.mod requirement for %q: %w", orchestrionPkgPath, err)
+	}
+
+	if required == version.Tag {
+		// This is the correct version or no specific version could be determined (indicating a dev/replaced package is in
+		// use), so we can proceed without further ado.
+		return nil
+	}
+
+	if respawn := osGetenv(envVarRespawned); respawn != "" && respawn != envValRespawnLocal {
+		// We're already re-spawning for a non-local version, so we should not be re-spawning again...
+		// If that were the case, we'd likely end up in an infinite loop of re-spawning, which is very
+		// much undesirable.
+		return fmt.Errorf(
+			"%w (wanted %s, got %s, already respawning for %s)",
+			errRespawnLoop,
+			required,
+			version.Tag,
+			respawn,
+		)
+	}
+
+	if required == "" {
+		// If there is no required version, it means a local version is used instead, either because we
+		// are in Orchestrion's own development tree, or because the user has introduced a "replace"
+		// directive for orchestion. In such cases, we unconditionally exec `go run` exactly once.
+		required = envValRespawnLocal
+	}
+
+	log.Printf("Re-starting with '%s@%s' (this is %s)\n", orchestrionPkgPath, required, version.Tag)
+
+	goBin, err := exec.LookPath("go")
+	if err != nil {
+		return fmt.Errorf("failed to resolve go from PATH: %w", err)
+	}
+
+	args := make([]string, len(osArgs)+2)
+	args[0] = goBin
+	args[1] = "run"
+	args[2] = orchestrionPkgPath
+	copy(args[3:], osArgs[1:])
```

Comment:
This assumes that `requiredVersion` gets called with len(osArgs) >= 1, otherwise we're writing out of bounds. Since this seems to be a strong assertion given that we expect to never receive an empty slice (probably copied from `os.Args`), why not assert it in the code with a panic?